[WIP] Add RCH integration to pull data from RCH API and list out

mkangia opened the pull request.

mkangia committed:
Add RCH integration to pull data from RCH API and list out [skip ci]

dimagimon commented:
There are 5 errors:

* custom/rch/forms.py, line 6 - F401 'Fieldset' imported but unused
* custom/rch/models.py, line 139 - E226 missing whitespace around arithmetic operator
* custom/rch/models.py, line 187 - E226 missing whitespace around arithmetic operator
* custom/rch/utils.py, line 5 - E203 whitespace before ':'
* custom/rch/views.py, line 6 - F401 'AreaMapping' imported but unused

snopoke commented on custom/rch/forms.py:
```diff
@@ -0,0 +1,109 @@
+
+
+def get_state_names_choices():
+    states = AreaMapping.objects.order_by().values_list('stname', flat=True).distinct()
```
what does `order_by()` do?

mkangia replied:
I thought this was needed here. Cant remember why. Removing it.

snopoke commented on custom/rch/forms.py:
```diff
@@ -0,0 +1,109 @@
+    options = set()
+    for district in districts:
+        options.add((district, district))
+    return tuple(options)
```
you could do all of this in a helper function and just pass in the field you want to query

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+    Religion = models.CharField(null=True, max_length=255)
+    State_Name = models.CharField(null=True, max_length=255)
+    StateID = models.PositiveSmallIntegerField(null=True)
+    Updated_On = models.DateTimeField(null=True)
```
Is there a reason to make these camel case? Normally DB columns should be lower case (makes querying easier since you don't have to quote column names)

mkangia replied:
The field names were kept same as the property names received intentionally so that https://github.com/dimagi/commcare-hq/pull/14554#discussion-diff-96423237R146 works. Ideally we should keep this camel case but then we would need a one to one mapping for these with the properties received from API. It would cost us some computation.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+    pjname = models.CharField(max_length=255, null=False)
+    awcid = models.BigIntegerField(null=False)
+    awcname = models.CharField(max_length=255, null=False)
+    villcode = models.IntegerField(null=False)
```
why `villcode` and not `village_code`? The second option is clearer and consistent with `village_name`

mkangia replied:
Its just to keep it same as the fields received from the API as shared [above](https://github.com/dimagi/commcare-hq/pull/14554#discussion-diff-96421945R32)

snopoke replied:
:+1:

snopoke replied:
:+1:

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+        try:
+            if self.Delivery_Time:
+                datetime.datetime.strptime(self.Delivery_Time, '%H:%M:%S')
+        except ValueError:
```
Why are these necessary? Doesn't the field do it's own validation? Custom validation should be done in the `clean` method: https://docs.djangoproject.com/en/1.8/ref/models/instances/#validating-objects

mkangia replied:
This is to take care if Delivery_Time is not in correct format. There was once the API returned "%H:%M" (AS-IS) in this field. So this takes care of it so that save call does not fail. I don't think we want the save to fail hence not validating it but just ignoring it if not in required format. Also read that clean method is not called on save(). Should it be called explicitly then?

snopoke replied:
I think it would be better to put this in a separate function which gets called in `update_beneficiaries` prior to calling save.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+
+        if self.Maternal_Death:
+            self.Maternal_Death = 0
+            if self.Maternal_Death == 'Y' or self.Maternal_Death == 'y':
```
This conversion should happen before the value gets assigned to the field.

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+                for record in records:
+                    rch_mother = cls()
+                    for prop in record:
+                        rch_mother.__setattr__(prop.keys()[0], prop.values()[0])
```
rather use `setattr(rch_mother, key, value)`

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+                    rch_child = cls()
+                    for prop in record:
+                        rch_child.__setattr__(prop.keys()[0], prop.values()[0])
+                    rch_child.save()
```
seems very similar to method on `RCHMother`, could extract a helper function?

mkangia replied:
I feel models would be right place and not helper function since these are actually adding records. Moved to parent class if that sounds right.

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,88 @@
+        context = self.get_context_data()
+        mother_beneficiaries = RCHMother.objects
+        child_beneficiaries = RCHChild.objects
+        state = request.POST['state']
```
should use `POST.get` to avoid key errors if the parameter is missing

snopoke commented on requirements/requirements.txt:
```diff
@@ -97,3 +97,4 @@ csiphash==0.0.4
 json-delta==2.0
 email_validator==1.0.1
 dnspython==1.13.0
+git+https://github.com/mvantellingen/python-zeep.git@46653fac2a92a99bc93d20c772cd7d797c301105
```
any reason not to use the version on pypi: https://pypi.python.org/pypi/zeep?

mkangia replied:
This took quite a while, to figure out the version that would work.  This commit version is just above one stable version and this particular [commit](https://github.com/mvantellingen/python-zeep/commit/46653fac2a92a99bc93d20c772cd7d797c301105) is a fix for empty tags. The recent version failed due to some parsing errors.

snopoke commented on settings.py:
```diff
@@ -370,6 +371,10 @@
 
+RCH_APPS = (
+    'custom.rch',
+)
```
you don't need this if you've already added it to `HQ_APPS`

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,212 @@
+    Weight = models.FloatField(null=True)
+
+    @classmethod
+    def update_beneficiaries(cls):
```
Where do these get called?

mkangia replied:
These need to get bind to task decorators to be run everyday only between 6-8PM IST. Need to figure that out.

snopoke commented on custom/rch/views.py:
```diff
@@ -0,0 +1,88 @@
+from custom.rch.models import RCHMother, AreaMapping, RCHChild
+
+
+class BeneficariesList(TemplateView):
```
Are these views just for debugging purposes?

mkangia replied:
Yup! just views to list down Beneficiaries to show what data has been received from API.

mkangia committed:
extract a helper method for field choices

mkangia committed:
extract updating beneficiaries to parent class instead

mkangia committed:
use get to fetch value

mkangia committed:
require super access to view beneficiaries and beneficiary details

mkangia committed:
add periodic tasks for fetching beneficiaries at 18:00 and 18:15

mkangia committed:
remove unnecessary settings as custom.rch already in HQ_APPS

mkangia committed:
Add MDDS_VillageID to RCHRecord instead of just RCHChild

mkangia committed:
fix time for periodic task to be in PM [skip ci]

dimagimon commented on custom/rch/forms.py:
```diff
@@ -0,0 +1,77 @@
+from django.utils.translation import ugettext_lazy
+
+from crispy_forms import layout as crispy
+from crispy_forms.layout import Layout, Fieldset, ButtonHolder, Submit
```
F401 'Fieldset' imported but unused

dimagimon commented on custom/rch/utils.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def etree_to_dict(t):
+    return {t.tag : map(etree_to_dict, t.iterchildren()) or t.text}
```
E203 whitespace before ':'

dimagimon commented on custom/rch/views.py:
```diff
@@ -0,0 +1,97 @@
+
+from corehq.apps.domain.decorators import require_superuser
+from custom.rch.forms import BeneficiariesFilterForm
+from custom.rch.models import RCHMother, AreaMapping, RCHChild
```
F401 'AreaMapping' imported but unused

mkangia committed:
filter using MDDS_VillageID instead of Village_ID

mkangia committed:
fix dimagimon complaints [skip ci]

mkangia commented on custom/rch/utils.py:
```diff
@@ -0,0 +1,38 @@
+
+PROJECTID = ''  # From some secure place
+ID = ''  # From some secure place
+PASSWORD = ''  # From some secure place
```
@snopoke How should this be done? Adding secure ids and password.

mkangia committed:
Add clear button that clears filter and ugettext_lazy -> _

mkangia committed:
move filter submit as a GET request and include total records in the view as well until pagination integrated

Move filter submit to GET to have params in url and avoid form submission request on browser when navigating back after searching and clicking for beneficiary's details

mkangia committed:
Include an empty column for CAS fields

This would help compare values between two systems, since CAS is not available just include N/A

mkangia committed:
update to use GET for filtering by village ids when included

mkangia committed:
minor styling improvements [skip ci]

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,207 @@
+    Weight = models.FloatField(null=True)
+
+    @classmethod
+    def update_beneficiaries(cls, beneficiary_type=CHILD_DATA_TYPE):
```
probably don't need the `beneficiary_type` kwarg since I doubt you'll ever call it with something different.

mkangia replied:
Agreed. Added this to match the signature in parent class which is called later. Should that be a good practice?

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,207 @@
+                for record in records:
+                    rch_beneficiary = cls()
+                    for prop in record:
+                        setattr(rch_beneficiary, prop.keys()[0], prop.values()[0])
```
To make this a bit safer you could get all the fields on the model and make sure the props are a valid field. You could also do validation with it if you want:
```python
valid_fields = {field.name: field for field in cls._meta.fields}
field = prop.keys()[0]
if field not in valid_fields:
    # skip or do something
if isinstance(valid_fields[field], models.TimeField):
    # check that it is correct format
```

mkangia committed:
Add email notifiers for Unavailable service, insecure access response and parsing error

mkangia committed:
Rescue and notify in case Service Unavailable, Response Parsing error or insecure access in case of no records [skip ci]

mkangia committed:
update existing beneficiaries using Registration as pk for RCHMother and Child_RCH_ID_No for RCHChild [skip ci]

dimagimon commented on custom/rch/emails.py:
```diff
@@ -0,0 +1,26 @@
+    subject = "[RCHServiceParsingError] [{for_date}] Issue with parsing RCH response".format(for_date=for_date)
+    content = ("Unexpected response received for date:{for_date}, state_id:{s_id}, beneficiary_type:{bt}, "
+               "district_id: {d_id}".format(
+                for_date=for_date, s_id=state_id, bt=beneficiary_type, d_id=district_id))
```
E121 continuation line under-indented for hanging indent

dimagimon commented on custom/rch/emails.py:
```diff
@@ -0,0 +1,26 @@
+    content = ("Insecure access response received for date:{for_date}, state_id:{s_id}, beneficiary_type:{bt}, "
+               "district_id: {d_id}. Probably no updates in RCH for the requested date. Must check this if "
+               "consecutively happening.".format(
+                for_date=for_date, s_id=state_id, bt=beneficiary_type, d_id=district_id))
```
E121 continuation line under-indented for hanging indent

mkangia committed:
fix dimagimon indentation complaints

dimagimon commented on custom/rch/emails.py:
```diff
@@ -0,0 +1,26 @@
+    subject = "[RCHServiceParsingError] [{for_date}] Issue with parsing RCH response".format(for_date=for_date)
+    content = ("Unexpected response received for date:{for_date}, state_id:{s_id}, beneficiary_type:{bt}, "
+               "district_id: {d_id}".format(
+                    for_date=for_date, s_id=state_id, bt=beneficiary_type, d_id=district_id))
```
E126 continuation line over-indented for hanging indent

dimagimon commented on custom/rch/emails.py:
```diff
@@ -0,0 +1,26 @@
+    content = ("Insecure access response received for date:{for_date}, state_id:{s_id}, beneficiary_type:{bt}, "
+               "district_id: {d_id}. Probably no updates in RCH for the requested date. Must check this if "
+               "consecutively happening.".format(
+                    for_date=for_date, s_id=state_id, bt=beneficiary_type, d_id=district_id))
```
E126 continuation line over-indented for hanging indent

snopoke commented on custom/rch/emails.py:
```diff
@@ -0,0 +1,26 @@
+def notify_error_in_service(message, for_date):
+    subject = "[RCHServiceDown] [{for_date}] Unable to access RCH Service".format(for_date=for_date)
+    content = "Error in RCH Service: {message}".format(message=message)
+    send_mail_async.delay(subject, content, settings.DEFAULT_FROM_EMAIL, ['mkangia@dimagi.com'])
```
Can you not just use `notify_exception` for these?
```python
notify_exception(None, 'RCH Insecure access response received', {'date': for_date, 'state_id': state_id, 'beneficiary_type': bt, 'district_id': d_id})
```

mkangia replied:
Added a separate notification and only to me as of now because:
1. Not useful to all devs since its specific to RCH
2. Should not get missed in exception emails since this should be looked at asap considering we can fetch details only for last 7 days.

mkangia committed:
Fetch Credentials from settings and raise Error in case request initiated without adding them [skip ci]

mkangia commented on custom/rch/views.py:
```diff
@@ -0,0 +1,93 @@
+from custom.rch.models import RCHMother, RCHChild
+
+
+class BeneficariesList(TemplateView):
```
There would be quite some customisation to this. So would be better to keep it outside admin. Plus might be used by some non-admins as well, later.

snopoke commented:
@mkangia let me know when I should review this

mkangia commented:
@snopoke running late on a couple of changes. Also the deploy for this has been delayed due to permission issues for deploy. Getting that aligned as well.

mkangia committed:
update acceptable fields to be in son for usability, corrected couple of names and added missing fields

mkangia committed:
sanitise and accept only allowed fields, allow to pass number of days prior data to be fetched [skip ci]

mkangia committed:
update village id, name fields [skip ci]

snopoke commented on custom/rch/models.py:
```diff
@@ -0,0 +1,240 @@
+        date_str = str(datetime.date.fromordinal(datetime.date.today().toordinal() - days_before))
+        for state_id in STATE_DISTRICT_MAPPING:
+            for district_id in STATE_DISTRICT_MAPPING[state_id]:
+                records = fetch_beneficiaries_records(date_str, date_str, state_id, beneficiary_type, district_id)
```
are `from_date` and `to_date` meant to be the same?

mkangia replied:
yup. Could pass just one day as well. Will update

mkangia committed:
add all fields being received from api

mkangia committed:
Merge master: resolve conflict

mkangia committed:
add filter for source of beneficiaries

dimagimon commented on custom/rch/views.py:
```diff
@@ -0,0 +1,108 @@
+        context['child_beneficiaries'] = self.child_beneficiaries.order_by()[:20]
+
+
+        return context
```
E303 too many blank lines (2)

mkangia committed:
delete all migrations to make a single one

mkangia committed:
update form for better naming of variables

mkangia committed:
just retain a single model for RCH Records and keep details in Couch Doc

mkangia committed:
update views to reflect new changes

mkangia committed:
update management command to create mappings for new django version

mkangia committed:
default to show some rch records

dimagimon commented on custom/rch/models.py:
```diff
@@ -0,0 +1,146 @@
+    aadhar_num = models.BigIntegerField(null=True)
+    dob = models.DateTimeField(null=True)
+    rch_id = models.BigIntegerField(null=True)
+    doc_type = models.CharField(max_length=1, choices=[('0', 'mother'), ('1','child')], null=False)
```
E231 missing whitespace after ','

mkangia committed:
villcode -> village_code, Village_name -> village_name

mkangia committed:
remove redundant variable

mkangia committed:
update celery tasks to pull data according to new model structure

mkangia committed:
fix dimagimon complaint

mkangia commented:
@snopoke 
The process of taking data from RCH application needs to be reconsidered since we can't pull data from RCH like we have been assuming till now. :(
This came from our meeting at UIDAI for using Aadhar number as the common key for beneficiary matching. This might be needed later. Will revive it then.

i am going to close this pull request without merging it.

snopoke commented:
OK, I'm keen to chat about this more to understand what's changed.
